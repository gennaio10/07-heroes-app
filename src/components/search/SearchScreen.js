import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const initialForm = {
    searchText: q,
  };
  const [formValues, handleInputChange] = useForm(initialForm);
  const { searchText } = formValues;
  const heroesFilter = useMemo(() => getHeroesByName(q), [q]);

  const handleSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };

  return (
    <div className="container">
      <h1>SearchScreen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />

          <form onSubmit={handleSubmit}>
            <input
              type="text"
              name="searchText"
              placeholder="find your hero"
              className="form-control"
              value={searchText}
              onChange={handleInputChange}
            />

            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col">
          <h4>Results</h4>
          <hr />

          {q === "" && <div id="divBuscaAlgo" className="alert alert-info">Search a hero</div>}

          {q !== "" && heroesFilter.length === 0 && (
            <div id="divNoExiste" className="alert alert-danger">
              There is no a hero with {q}
            </div>
          )}

          {heroesFilter.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};
