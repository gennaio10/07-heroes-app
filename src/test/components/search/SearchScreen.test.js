import React from "react";
import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("Pruebas de <SearchScreen />)", () => {
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route exact path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("button").text().trim()).toBe("Search...");
  });

  test("debe de mostrar batman y le input con el valor sel querystring", () => {
    const searchText = "batman";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${searchText}`]}>
        <Route exact path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("input").prop("value")).toBe(searchText);
  });

  test("debe de mostrar la alerta de error si no se encuentra el hero", () => {
    const searchText = "batman-no-existe";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${searchText}`]}>
        <Route exact path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("#divNoExiste").exists()).toBe(true);
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      `There is no a hero with ${searchText}`
    );
  });

  test("debe de llamar el push del histori", () => {
    const searchText = "batman";

    const historyMock = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${searchText}`]}>
        <Route
          exact
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        value: searchText,
        name: "searchText",
      },
    });

    const formSubmit = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    expect(historyMock.push).toHaveBeenCalledTimes(1);
    expect(historyMock.push).toHaveBeenCalledWith(`?q=${searchText}`);
  });
});
