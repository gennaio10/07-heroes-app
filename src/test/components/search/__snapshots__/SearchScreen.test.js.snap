// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <SearchScreen />) debe de mostrar batman y le input con el valor sel querystring 1`] = `
<div
  className="container"
>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": 300,
          }
        }
      >
        <img
          alt="Batman"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="col-md-8"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Detective Comics #27
              </small>
            </p>
            <a
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas de <SearchScreen />) debe de mostrar la alerta de error si no se encuentra el hero 1`] = `
<div
  className="container"
>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value="batman-no-existe"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
        id="divNoExiste"
      >
        There is no a hero with 
        batman-no-existe
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas de <SearchScreen />) debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="container"
>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
        id="divBuscaAlgo"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
